assets/hdfs_reader/index.js

```diff
@@ -0,0 +1,239 @@
+    const opConfig = executionContext.config.operations[0];
+    const clientService = getClient(context, opConfig, 'hdfs_ha');
+    const hdfsClient = clientService.client;
+    const queue = new Queue();
+
+    function processFile(file, filePath) {
+        const totalLength = file.length;
+        let fileSize = file.length;
+
+        if (fileSize <= opConfig.size) {
+            queue.enqueue({ path: `${filePath}/${file.pathSuffix}`, fullChunk: true });
+        } else {
+            let offset = 0;
+            while (fileSize > 0) {
+                const length = fileSize > opConfig.size ? opConfig.size : fileSize;
+                // This grabs the character immediately before the chunk to help check whether or
+                // not the chunk starts with a complete record
+                let startSpot = offset;
+                if (offset > 0) {
+                    startSpot = offset - 1;
+                }
+                queue.enqueue({
+                    path: `${filePath}/${file.pathSuffix}`,
+                    offset: startSpot,
+                    length,
+                    total: totalLength
+                });
+                fileSize -= opConfig.size;
+                offset += length;
+            }
+        }
+    }
+
+    function getFilePaths(filePath) {
+        return hdfsClient.listStatusAsync(filePath)
+            .then(results => Promise.map(results, (metadata) => {
+                if (metadata.type === 'FILE') {
+                    return processFile(metadata, filePath);
+                }
+                if (metadata.type === 'DIRECTORY') {
+                    return getFilePaths(`${filePath}/${metadata.pathSuffix}`);
+                }
+                return true;
+            }))
+            .return([() => queue.dequeue()])
+            .catch(err => Promise.reject(parseError(err)));
+    }
+
+    return getFilePaths(opConfig.path)
+        .catch((err) => {
+            const errMsg = parseError(err);
+            logger.error(`Error while reading from hdfs, error: ${errMsg}`);
+            return Promise.reject(errMsg);
+        });
+}
+
+
+function newReader(context, opConfig) {
+    const formatters = {
+        jsonLines: (data, logger) => data.map(record => JSON.parse(record))
+            .filter(element => element !== undefined).catch((err) => {
+                logger.error(`There was an error processing the record: ${err}`);
+            }),
+    };
+    // Set up a logger
+    const logger = context.apis.foundation.makeLogger({ module: 'hdfs_reader' });
+    const clientService = getClient(context, opConfig, 'hdfs_ha');
+    const hdfsClient = clientService.client;
+    const chunkFormater = formatters[opConfig.format];
+
+    return function processSlice(slice) {
+        // Hardcoding the line delimiter, but this could be arbitrary
+        // TODO paramaterize this in the config
+        const lineDelimiter = '\n';
+        return getChunk(hdfsClient, slice, lineDelimiter)
```

Comment:
This is the start of the logic that I'm thinking can be pulled out of this reader and made into a generic library that can work on both HDFS and other file systems. It will need some thought on how to abstract the file system requests but I think it should be doable and then hopefully both HDFS reader and file system readers can gain the same set of capabilities as the library is improved..